### DISCUSSION TITLE
Generating PDEContainer groups

### DISCUSSION THREAD

**Discussion by Andreasox** [pymupdf/PyMuPDF#3049] [Q&A]
When creating PDF from e.g. QGIS, curved texts are generated as one object for each character.
For more efficient handling in e.g. Illustrator, we are planning to use the PDEContainer concept by inserting BMC/EMC flags before/after each text group. To achieve similar things we have inserted special characters as flags and replaced them in EPS/PDF files to desired PDF operators. Is something similar possible with pymupdf?

In the attached image, two PDEContainers should be created, "körväg stig" and "försvunnen" for the current 22 character objects, see attached PDF. We may generate the text as e.g. "^försvunnen^" where ^ should be replaced with BMC/EMC
<img width="278" alt="image" src="https://github.com/pymupdf/PyMuPDF/assets/6737097/b1701dfd-6150-4d2f-981d-5bea7af422a9">
 
[qgis.pdf](https://github.com/pymupdf/PyMuPDF/files/13827933/qgis.pdf)

**Comment by Andreasox** [comment]
https://www.citywayfinding.com/ shows what we are producing in reality.

**Comment by JorjMcKie** [comment]
I suggest we make a Discussion post for this. There are ways to edit a page's `/Contents` and one of the several text output options also directly edits these objects.

Maybe something can be made from this constellation. For a major change resulting from this corner case request we would certainly see a business case for us.

So let's see how far a this discussion will take us.

**Comment by Andreasox** [comment]
T-Kartor will internally investigate if PDEContainers is the same thing as Illustrator groups, or can be converted using the Illustrator SDK. Our primary use case is to facilitate editing in Illustrator.

**Comment by Andreasox** [comment]
We are quite certain that PDEContainers (BDC/BMC-EMC tag pairs) are correctly read by Illustrator, according to syntax from https://opensource.adobe.com/dc-acrobat-sdk-docs/pdflsdk/apireference/PDFEdit_Layer/PDEContainer.html
We will test further and report any issues which may have relevance to an implementation in pyMuPDF.
Attaching an uncompressed PDF file (readable in any text editor) where examples can be found.
[ocad_minimal.pdf](https://github.com/pymupdf/PyMuPDF/files/13960363/ocad_minimal.pdf)

**Reply by JorjMcKie** [reply]
I have looked at the example PDF.
At first sight I must say that this represents an implementation effort significantly beyond what we will take into consideration in the foreseeable future - at least for the GNU AGPL license.

**Comment by Andreasox** [comment]
Understand. I will come back if licensing/resources needs to be discussed. We are still in an investigating phase to select softwares and to estimate our development efforts according to available resources.

**Reply by JorjMcKie** [reply]
Thank you very much for your understanding.
Please do not hesitate to [contact](https://artifex.com/contact/pymupdf-inquiry.php) us to discuss a suitable solution.
